### PR TITLE
Add datamodel changes for secret support to Terraform Providers 

### LINKED ISSUE #6539: Allow any Terraform provider to be configured and used in a Recipe
## Overview of feature request

Today we support Terraform templates for Recipes. For the Azure, Kubernetes, and AWS providers we automatically setup the provider configuration with the identity of where to deploy and we pass in where to deploy via the `context` parameter.

Additionally, any additional provider can be used in a Recipe, however there is no way to securely configure the provider with credentials. For example, there is no way to use the GCP Terraform Provider in a Recipe without hard-coding the credentials in the Recipe module.

This feature is for allowing credentials to be securely configured and passed into the Recipe so any Terraform Provider can be used.

## Acceptance criteria

Once this feature is implemented I should be able to use the GCP, Oracle Cloud, or any other Terraform provider and pass in the credentials. I should be able to follow documentation that walks me through the steps I should take to accomplish this, with snippets showing an example.

There should be unit tests confirming that credentials are passed in correctly.

## Additional context

There are a couple ways to implement this, as each Terraform provider handles auth in various ways (_managed identities, service principals, IAM roles, etc._). We should take a look at how most providers handle credentials and see what makes sense to implement to cast the widest nest. Could be parameters that are passed in with the credentials, or could be a Terraform RC file pass through.

## Community ask

Please comment below with any Terraform modules that you would like to use.

[AB#9883](https://dev.azure.com/azure-octo/e61041b4-555f-47ae-95b2-4f8ab480ea57/_workitems/edit/9883)

### PR DESCRIPTION
**Pull request by lakshmimsft**
# Description

Added datamodel updates for secret support to Terraform Providers configuration. This includes secrets in recipeConfig under specific Provider configurations and environment variables. 
Pls see link to associated [Design Doc](https://github.com/radius-project/design-notes/pull/47/files)

NOTE: Updates to functional tests are commented till we merge in corresponding PR in bicep repo : https://github.com/radius-project/bicep/pull/751

## Type of change
- This pull request fixes a bug in Radius and has an approved issue (#6539)

Fixes: #6539

### CODE REVIEW

**Review by kachawla — COMMENTED** [review]
@lakshmimsft could you update the description with link to the API design doc corresponding to these changes?

**Review by kachawla — COMMENTED** [review]
@lakshmimsft are typespec changes mostly pulled in from https://github.com/radius-project/radius/pull/7199 or have they diverged significantly since then?

File: typespec/Applications.Core/environments.tsp
Comment (kachawla) [inline typespec/Applications.Core/environments.tsp:121]: What is "referenced resource" in this context?

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (kachawla) [inline pkg/corerp/api/v20231001preview/environment_conversion.go:433]: we don't really abbreviate `additional` to `addnl` anywhere else in the code base

```suggestion
			var additionalProperties map[string]any
```

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (kachawla) [inline pkg/corerp/api/v20231001preview/environment_conversion.go:434]: converted from what? Can we just call it `tfProviderSecrets`?

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (kachawla) [inline pkg/corerp/api/v20231001preview/environment_conversion.go:437]:
```suggestion
			providerProperties = providerAdditionalProperties.AdditionalProperties
```

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (kachawla) [inline pkg/corerp/api/v20231001preview/environment_conversion.go:441]: Do we expect users to add additional properties within each provider config?

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (kachawla) [inline pkg/corerp/api/v20231001preview/environment_conversion.go:462]: This code is getting difficult to follow with abbreviations being used which I'm sure are obvious to you while writing the code but I don't know how to interpret `tmpSecret` and `propSecret`. Can you look into making it more readable?

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (kachawla) [inline pkg/corerp/api/v20231001preview/environment_conversion.go:452]: Why are we fetching it from [additionalProperties](https://github.com/radius-project/radius/blob/3cd05edd9140bdc33a6487cf8516f117f8c6a354/pkg/corerp/api/v20231001preview/zz_generated_models.go#L1255) instead of this [explicit field assigned for secrets](https://github.com/radius-project/radius/blob/3cd05edd9140bdc33a6487cf8516f117f8c6a354/pkg/corerp/api/v20231001preview/zz_generated_models.go#L1259)?

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (kachawla) [inline pkg/corerp/api/v20231001preview/environment_conversion.go:590]: Why is this check needed?

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (kachawla) [inline pkg/corerp/api/v20231001preview/environment_conversion.go:515]: what is input and output type?

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (kachawla) [inline pkg/corerp/api/v20231001preview/environment_conversion.go:529]:
```suggestion
			providerConfig := ProviderConfigProperties{
```

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (kachawla) [inline pkg/corerp/api/v20231001preview/environment_conversion.go:522]: Wouldn't it be always nil at this point?

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (kachawla) [inline pkg/corerp/api/v20231001preview/environment_conversion.go:520]: anything preventing us from directly assigning it to `convertedSecrets` here instead of creating an intermediate tmp variable?

```suggestion
				convertedSecrets[secretKey] := &SecretReference{
					Source: to.Ptr(secretValue.Source),
					Key:    to.Ptr(secretValue.Key),
				}
```

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (kachawla) [inline pkg/corerp/api/v20231001preview/environment_conversion.go:513]: we are using `providers` above and not `convertedProviders`, can we use same convention here and just name it `secrets`?

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (kachawla) [inline pkg/corerp/api/v20231001preview/environment_conversion.go:483]: We have a specific convention for function names for to/from conversions between internal and external data models, can you update this function name to follow that?

Comment (kachawla) [inline pkg/corerp/api/v20231001preview/environment_conversion.go:483]:
```suggestion
func getSecretsFromProviderProperties(providerSecrets map[string]*SecretReference) map[string]datamodel.SecretReference {
```

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (ytimocin) [inline pkg/corerp/api/v20231001preview/environment_conversion.go:208]:
```suggestion
			recipeConfig.Terraform.Providers = toRecipeConfigTerraformProvidersDatamodel(config)
```

Is there a reason why we don't do it like above?

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (ytimocin) [inline pkg/corerp/api/v20231001preview/environment_conversion.go:212]: This is in-lined and the other one is not.

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (ytimocin) [inline pkg/corerp/api/v20231001preview/environment_conversion.go:433]: We can inline this part too.

```suggestion
				AdditionalProperties: providerConfigProperties.AdditionalProperties,
				Secrets:              getSecretsFromProviderProperties(providerConfigProperties.Secrets),
```

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (ytimocin) [reply pkg/corerp/api/v20231001preview/environment_conversion.go:473]: Can this be deleted?

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (ytimocin) [inline pkg/corerp/api/v20231001preview/environment_conversion.go:492]: `&providerConfig` can be in-lined too.

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (ytimocin) [inline pkg/corerp/api/v20231001preview/environment_conversion.go:552]:
```suggestion
		if envSecrets == nil {
			envSecrets = map[string]*SecretReference{}
		}

		envSecrets[k] = &SecretReference{
			Source: to.Ptr(v.Source),
			Key:    to.Ptr(v.Key),
		}
```

File: pkg/corerp/api/v20231001preview/environment_conversion_test.go
Comment (ytimocin) [inline pkg/corerp/api/v20231001preview/environment_conversion_test.go:991]: The test name should be updated if the function name gets updated.

File: test/functional-portable/corerp/noncloud/resources/recipe_terraform_test.go
Comment (ytimocin) [inline test/functional-portable/corerp/noncloud/resources/recipe_terraform_test.go:148]: Are these to be deleted?

File: test/functional-portable/corerp/noncloud/resources/recipe_terraform_test.go
Comment (ytimocin) [inline test/functional-portable/corerp/noncloud/resources/recipe_terraform_test.go:180]: Are these to be deleted?

File: test/functional-portable/corerp/noncloud/resources/recipe_terraform_test.go
Comment (ytimocin) [inline test/functional-portable/corerp/noncloud/resources/recipe_terraform_test.go:190]: Is this to be deleted?

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (lakshmimsft) [reply pkg/corerp/api/v20231001preview/environment_conversion.go]: i was in middle of updating files. updated now.

File: typespec/Applications.Core/environments.tsp
Comment (lakshmimsft) [reply typespec/Applications.Core/environments.tsp]: updated comment to "Sensitive data in provider configuration can be stored as secrets..."

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (lakshmimsft) [reply pkg/corerp/api/v20231001preview/environment_conversion.go]: function has been updated. pls take another look.

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (lakshmimsft) [reply pkg/corerp/api/v20231001preview/environment_conversion.go]: no.. this is not needed. function has been updated. pls take another look.

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (lakshmimsft) [reply pkg/corerp/api/v20231001preview/environment_conversion.go]: This has been removed. pls take another look.

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (lakshmimsft) [reply pkg/corerp/api/v20231001preview/environment_conversion.go]: This is not needed and has been removed.

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (lakshmimsft) [reply pkg/corerp/api/v20231001preview/environment_conversion.go:530]: this function has been refactored. the check is needed to ensure we instantiate only when needed

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (lakshmimsft) [reply pkg/corerp/api/v20231001preview/environment_conversion.go]: updated

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (lakshmimsft) [reply pkg/corerp/api/v20231001preview/environment_conversion.go]: updated.

File: pkg/corerp/api/v20231001preview/environment_conversion_test.go
Comment (lakshmimsft) [reply pkg/corerp/api/v20231001preview/environment_conversion_test.go]: updated

File: test/functional-portable/corerp/noncloud/resources/recipe_terraform_test.go
Comment (lakshmimsft) [reply test/functional-portable/corerp/noncloud/resources/recipe_terraform_test.go:180]: will enable these to test functional test updates once bicep PR is merged in.

File: pkg/corerp/api/v20231001preview/environment_conversion.go
Comment (lakshmimsft) [reply pkg/corerp/api/v20231001preview/environment_conversion.go]: function is refactored and names updated. pls take another look.

File: test/functional-portable/corerp/noncloud/resources/recipe_terraform_test.go
Comment (lakshmimsft) [reply test/functional-portable/corerp/noncloud/resources/recipe_terraform_test.go:190]: i will uncomment this to test functional test updates once bicep PR is merged in.

**Review by kachawla — COMMENTED** [review]
Looks good to go once the functional test is uncommented

File: test/functional-portable/corerp/noncloud/resources/testdata/corerp-resources-terraform-postgres.bicep
Comment (ytimocin) [inline test/functional-portable/corerp/noncloud/resources/testdata/corerp-resources-terraform-postgres.bicep:76]: may need a reformat here

**Review by ytimocin — DISMISSED** [review]

**Review by kachawla — APPROVED** [review]

**Pull request merged**
